templating/secrets.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package templating
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/acr-builder/graph"
+	"github.com/Azure/acr-builder/vaults"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/pkg/errors"
+)
+
+// ResolveSecretFunc is a function that resolves the secret to its value and sends through the ResolvedChan of the secret. Any errors during resolve are send through errorChan
+type ResolveSecretFunc func(ctx context.Context, azureVaultResourceURL string, secret *graph.Secret, errorChan chan error)
+
+// SecretResolver defines how a secret is resolved.
+type SecretResolver struct {
+	Resolve          ResolveSecretFunc
+	azureEnvironment azure.Environment
```

Comment:
SecretResolver sounds like a generic and not bound to azure. Maybe AzureSecretResolver?

Reply:
SecretResolver is generic and should be able abstract from the vault it is being resolved

Reply:
I see. If SecretResolver is designed to be generic interface for all providers, it probably should not include azureEnvironment. Ideally ResolveSecretFunc can abstract it. Or you can introduce another AzureSecretResolver based on SecretResolver. 
It's not critical. I think we can address it later.